### PR TITLE
wasm: Add Wasmtime as a new runtime

### PR DESCRIPTION
**Pull request by mathetake**
Signed-off-by: mathetake <takeshi@tetrate.io>

Commit Message: wasm: Add Wasmtime(https://github.com/bytecodealliance/wasmtime) as a new runtime for Wasm filter
Additional Description: add a new wasm runtime named Wasmtime
Risk Level: low
Testing: run in `bazel.compile_time_options` build target as WAVM.
Docs Changes: Docs will be provided in subsequent PRs.
Release Notes: enable Wasmtime as a Wasm filter runtime optionally.
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]

ref: https://github.com/proxy-wasm/proxy-wasm-cpp-host/pull/73

### CODE REVIEW

**Comment by antoniovicente** [comment]
Thanks for the pull request.  Is there some particular aspect of this change that you'ld like early feedback on?  What additional changes you think are needed before this draft PR is ready for review and merge?

File: source/extensions/common/wasm/well_known_names.h
Comment (PiotrSikora) [inline source/extensions/common/wasm/well_known_names.h:18]: Nit: `s/wasmtime/Wasmtime/`

File: bazel/external/proxy_wasm_cpp_host.BUILD
Comment (PiotrSikora) [reply bazel/external/proxy_wasm_cpp_host.BUILD:34]: Also, perhaps we should remove the "all" option altogether, since it no longer represents all runtimes?

Multiple runtimes can be selected individually anyway, e.g.:
```
bazel build --config=libc++ --define=wasm=v8 --define=wasm=wavm //source/exe:envoy-static
```

Comment (mathetake) [reply bazel/external/proxy_wasm_cpp_host.BUILD:34]: > Multiple runtimes can be selected individually anyway, e.g.:

I didn't know that. Thanks! I'll remove `all` option

Comment (mathetake) [reply bazel/external/proxy_wasm_cpp_host.BUILD:34]: and, let me see if the collision with WAVM has been resolved

File: ci/do_ci.sh
Comment (PiotrSikora) [inline ci/do_ci.sh:283]: We need to either enable both runtimes here (if the LLVM conflict is resolved), or move Wasmtime to another build target, since we need to have coverage for all runtimes in the CI.

File: bazel/external/proxy_wasm_cpp_host.BUILD
Comment (mathetake) [reply bazel/external/proxy_wasm_cpp_host.BUILD:34]: I could observe the same collision with `jit-dump=disabled`: 
```
se --sandbox_debug to see verbose messages from the sandbox
ld.lld: error: duplicate symbol: __jit_debug_descriptor
>>> defined at GDBRegistrationListener.cpp
>>>            GDBRegistrationListener.cpp.o:(__jit_debug_descriptor) in archive bazel-out/k8-fastbuild/bin/bazel/foreign_cc/llvm/lib/libLLVMExecutionEngine.a
>>> defined at wasmtime_runtime.849cvge1-cgu.15
>>>            wasmtime_runtime-361389515.wasmtime_runtime.849cvge1-cgu.15.rcgu.o:(.data.__jit_debug_descriptor+0x0) in archive bazel-out/k8-fastbuild/bin/external/com_github_wasmtime/librust_c_api-220117231.a

ld.lld: error: duplicate symbol: __jit_debug_register_code
>>> defined at GDBRegistrationListener.cpp
>>>            GDBRegistrationListener.cpp.o:(__jit_debug_register_code) in archive bazel-out/k8-fastbuild/bin/bazel/foreign_cc/llvm/lib/libLLVMExecutionEngine.a
>>> defined at wasmtime_runtime.849cvge1-cgu.15
>>>            wasmtime_runtime-361389515.wasmtime_runtime.849cvge1-cgu.15.rcgu.o:(.text.__jit_debug_register_code+0x0) in archive bazel-out/k8-fastbuild/bin/external/com_github_wasmtime/librust_c_api-220117231.a
```

It's defined in the runtime package which is nothing to dow with `jit-dump` feature https://github.com/bytecodealliance/wasmtime/blob/v0.21.0/crates/runtime/src/jit_int.rs#L32-L41

Comment (mathetake) [reply bazel/external/proxy_wasm_cpp_host.BUILD:34]: and I realized that the V8 defines the same symbol but disable it by default https://chromium.googlesource.com/v8/v8/+/4.10.71/src/gdb-jit.cc#24 , that's why luckily we have not encountered symbol collision in conjunction with V8 and WAVM

Comment (mathetake) [reply bazel/external/proxy_wasm_cpp_host.BUILD:34]: This ugly patch seems to work...

```
diff --git a/CMakeLists.txt b/CMakeLists.txt
index eb5d7c2c..2bd2c748 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -720,7 +720,6 @@ if(WAVM_ENABLE_RUNTIME)
 	add_subdirectory(Lib/Runtime)
 	add_subdirectory(Lib/ThreadTest)
 	add_subdirectory(Lib/WASI)
-	add_subdirectory(Lib/wavm-c)
 endif()
 
 if(WAVM_ENABLE_UNWIND)
diff --git a/Lib/LLVMJIT/LLVMModule.cpp b/Lib/LLVMJIT/LLVMModule.cpp
index 5f830959..0e222f36 100644
--- a/Lib/LLVMJIT/LLVMModule.cpp
+++ b/Lib/LLVMJIT/LLVMModule.cpp
@@ -61,9 +61,6 @@ using namespace WAVM::LLVMJIT;
 
 struct LLVMJIT::GlobalModuleState
 {
-	Platform::Mutex gdbRegistrationListenerMutex;
-	llvm::JITEventListener* gdbRegistrationListener = nullptr;
-
 	// A map from address to loaded JIT symbols.
 	Platform::RWMutex addressToModuleMapMutex;
 	std::map<Uptr, LLVMJIT::Module*> addressToModuleMap;
@@ -76,11 +73,8 @@ struct LLVMJIT::GlobalModuleState
 
 	// These constructor and destructor should not be called directly, but must be public in order
 	// to be accessible by std::make_shared.
-	GlobalModuleState()
-	{
-		gdbRegistrationListener = llvm::JITEventListener::createGDBRegistrationListener();
-	}
-	~GlobalModuleState() { delete gdbRegistrationListener; }
+	GlobalModuleState() {}
+	~GlobalModuleState() {}
 };
 
 // Allocates memory for the LLVM object loader.
@@ -516,13 +510,6 @@ Module::Module(const std::vector<U8>& objectBytes,
 
 	// Notify GDB of the new object.
 	{
-		Platform::Mutex::Lock lock(globalModuleState->gdbRegistrationListenerMutex);
-#if LLVM_VERSION_MAJOR >= 8
-		globalModuleState->gdbRegistrationListener->notifyObjectLoaded(
-			reinterpret_cast<Uptr>(this), *object, *loadedObject);
-#else
-		globalModuleState->gdbRegistrationListener->NotifyObjectEmitted(*object, *loadedObject);
-#endif
 	}
 
 	// Create a DWARF context to interpret the debug information in this compilation unit.
@@ -616,15 +603,7 @@ Module::Module(const std::vector<U8>& objectBytes,
 Module::~Module()
 {
 	// Notify GDB that the object is being unloaded.
-	{
-		Platform::Mutex::Lock lock(globalModuleState->gdbRegistrationListenerMutex);
-#if LLVM_VERSION_MAJOR >= 8
-		globalModuleState->gdbRegistrationListener->notifyFreeingObject(
-			reinterpret_cast<Uptr>(this));
-#else
-		globalModuleState->gdbRegistrationListener->NotifyFreeingObject(*object);
-#endif
-	}
+	{}
 
 	// Remove the module from the global address to module map.
 	{
diff --git a/Programs/wavm/CMakeLists.txt b/Programs/wavm/CMakeLists.txt
index 3d00c1df..e3445c13 100644
--- a/Programs/wavm/CMakeLists.txt
+++ b/Programs/wavm/CMakeLists.txt
@@ -13,7 +13,6 @@ set(NonRuntimeSources Testing/DumpTestModules.cpp
 set(RuntimeOnlySources
 			Testing/Benchmark.cpp
 			Testing/RunTestScript.cpp
-			Testing/TestCAPI.c
 			wavm-compile.cpp
 			wavm-run.cpp)
 
@@ -37,6 +36,3 @@ add_test(NAME HashMap COMMAND $<TARGET_FILE:wavm> test hashmap)
 add_test(NAME HashSet COMMAND $<TARGET_FILE:wavm> test hashset)
 add_test(NAME I128 COMMAND $<TARGET_FILE:wavm> test i128)
 
-if(WAVM_ENABLE_RUNTIME)
-	add_test(NAME C-API COMMAND $<TARGET_FILE:wavm> test c-api)
-endif()
diff --git a/Programs/wavm/Testing/wavm-test.cpp b/Programs/wavm/Testing/wavm-test.cpp
index eed47d13..61ef1189 100644
--- a/Programs/wavm/Testing/wavm-test.cpp
+++ b/Programs/wavm/Testing/wavm-test.cpp
@@ -99,7 +99,7 @@ int execTestCommand(int argc, char** argv)
 		case TestCommand::hashSet: return execHashSetTest(argc - 1, argv + 1);
 		case TestCommand::i128: return execI128Test(argc - 1, argv + 1);
 #if WAVM_ENABLE_RUNTIME
-		case TestCommand::cAPI: return execCAPITest(argc - 1, argv + 1);
+		case TestCommand::cAPI: return 0;
 		case TestCommand::benchmark: return execBenchmark(argc - 1, argv + 1);
 		case TestCommand::script: return execRunTestScript(argc - 1, argv + 1);
 #endif
```

Comment (mathetake) [reply bazel/external/proxy_wasm_cpp_host.BUILD:34]: with this patch, I could link both WAVM and Wasmtime at the same time, but when I run extension with wasmtime, I see weird error (I have completely not sure where this is coming from)
```
2020-11-10 15:11:32.316][576526][debug][wasm] [source/extensions/common/wasm/wasm.cc:126] Thread-Local Wasm created 3 now active
libunwind: _unw_add_dynamic_fde: bad fde: FDE is really a CIE
libunwind: _unw_add_dynamic_fde: bad fde: FDE is really a CIE
libunwind: _unw_add_dynamic_fde: bad fde: FDE is really a CIE
libunwind: _unw_add_dynamic_fde: bad fde: FDE is really a CIE
libunwind: _unw_add_dynamic_fde: bad fde: FDE is really a CIE
```

This is not observed when we link Wasmtime alone, so something's broken

Comment (PiotrSikora) [reply bazel/external/proxy_wasm_cpp_host.BUILD:34]: Yeah, I don't think we want to maintain that big of a patch. Removing single directory from `CMakeLists.txt` is acceptable, but modifying code to link multiple runtimes is probably not a good idea.

Thanks for looking, though!

File: ci/do_ci.sh
Comment (mathetake) [reply ci/do_ci.sh:283]: Given the discussion above, I don't think resolving symbol conflicts in this PR (only by leveraging a small patch) is feasible. So I will add another build for Wasmtime. After the merge, I'll file an issue for symbol collision.

Comment (mathetake) [reply ci/do_ci.sh:283]: @lizan Is there any existing build target I could set `--define wasm=wastime`, or should I add a new one for testing wasmtime?

**Comment by mathetake** [comment]
@antoniovicente sorry for the late response. The CI issue and symbol collision issue discussed above are supposed to be resolved before making this PR ready. Thanks for your help :smile:

File: bazel/repository_locations.bzl
Comment (PiotrSikora) [inline bazel/repository_locations.bzl:875]: You need to update `release_date` below to `2020-11-10`.

**Review by PiotrSikora — APPROVED** [review]
Thanks! LGTM once @lizan figures out the CI target we can use.

File: ci/do_ci.sh
Comment (lizan) [reply ci/do_ci.sh:283]: I think you can add a line below to do more test with wasmtime in same CI target. before we figure out #12574

File: bazel/external/proxy_wasm_cpp_host.BUILD
Comment (lizan) [inline bazel/external/proxy_wasm_cpp_host.BUILD:46]: I think you can still do glob, including extra header file shouldn't result symbol conflict.

Comment (mathetake) [reply bazel/external/proxy_wasm_cpp_host.BUILD]: done: https://github.com/envoyproxy/envoy/pull/13932/commits/b00c82a6198f2242dc98b7606ae754b56d1fed61

File: ci/do_ci.sh
Comment (mathetake) [reply ci/do_ci.sh]: https://github.com/envoyproxy/envoy/pull/13932/commits/ec4d2ddd33fefa7cee8508aaba7d499b968b0b85 done

File: test/extensions/filters/http/wasm/wasm_filter_test.cc
Comment (mathetake) [inline test/extensions/filters/http/wasm/wasm_filter_test.cc:1442]: this is a cherry pick from https://github.com/envoyproxy/envoy/pull/13930 so we want that PR to be merged first

File: ci/do_ci.sh
Comment (PiotrSikora) [inline ci/do_ci.sh:304]: You use `${COMPILE_TIME_OPTIONS}` twice, instead of `${BAZEL_BUILD_OPTIONS}`.

Also, I think the `--define wasm=wavm` is going to override `--define wasm=wasmtime` here. I'd prefer if you could remove `--define wasm=wavm` from `${COMPILE_TIME_OPTIONS}` and add `--define wasm=wavm` and `--define wasm=wasmtime` explicitly to each invocation to make it clearer.

File: bazel/repository_locations.bzl
Comment (PiotrSikora) [inline bazel/repository_locations.bzl:645]: This should be `2019-11-14`.

File: bazel/repository_locations.bzl
Comment (PiotrSikora) [inline bazel/repository_locations.bzl:626]: This should be `2020-11-05`.

Comment (mathetake) [reply bazel/repository_locations.bzl]: thanks!

File: bazel/repository_locations.bzl
Comment (mathetake) [reply bazel/repository_locations.bzl]: Thanks!

File: ci/do_ci.sh
Comment (mathetake) [reply ci/do_ci.sh]: thanks done in https://github.com/envoyproxy/envoy/pull/13932/commits/e4fa56ef895fff183abf84daf53a6d91227041bd

File: bazel/repository_locations.bzl
Comment (PiotrSikora) [inline bazel/repository_locations.bzl:864]: Nit: it looks that you accidentally updated release date of Proxy-Wasm C++ SDK instead of host.

**Review by PiotrSikora — APPROVED** [review]

File: bazel/repository_locations.bzl
Comment (mathetake) [reply bazel/repository_locations.bzl]: oh, thanks!

**Review by PiotrSikora — DISMISSED** [review]

File: ci/do_ci.sh
Comment (lizan) [inline ci/do_ci.sh:318]: I don't think we really need build binaries, here and above.

File: ci/do_ci.sh
Comment (lizan) [inline ci/do_ci.sh:303]: let's do wasmtime first, then the rest will be all wavm so it doesn't invalidate bazel cache.

Comment (mathetake) [reply ci/do_ci.sh:303]: right.

**Review by lizan — DISMISSED** [review]

**Comment by lizan** [comment]
Should be good to go but need @envoyproxy/dependency-shepherds review.

**Review by PiotrSikora — DISMISSED** [review]
@envoyproxy/dependency-shepherds friendly ping.

File: bazel/repository_locations.bzl
Comment (moderation) [inline bazel/repository_locations.bzl:645]: This commit is a year old. There have been 10 commits since with the latest on Nov 3 2020. Any reason why we are not using the latest?

File: bazel/repository_locations.bzl
Comment (moderation) [inline bazel/repository_locations.bzl:872]: Suggest updating to the latest commit
```
        version = "eceb02d5b7772ec1cd78a4d35356e57d2e6d59bb",
        sha256 = "ae9d9b87d21d95647ebda197d130b37bddc5c6ee3e6630909a231fd55fcc9069",
        release_date = "2020-11-10",
```

**Review by moderation — CHANGES_REQUESTED** [review]
Brings in two more deps - would be good to have @htuch review. Would like to bring in latest commits as per comments

File: bazel/repository_locations.bzl
Comment (PiotrSikora) [reply bazel/repository_locations.bzl:645]: This is included in Wasmtime as a submodule at the imported commit. There is no reason to diverge from that.

File: bazel/repository_locations.bzl
Comment (PiotrSikora) [reply bazel/repository_locations.bzl:872]: The latest commit requires a ton of other changes (see: https://github.com/envoyproxy/envoy/pull/13753).

File: bazel/repository_locations.bzl
Comment (moderation) [reply bazel/repository_locations.bzl:645]: Can't see where that specific commit is in https://github.com/bytecodealliance/wasmtime/blob/main/.gitmodules. If this old commit is the one we need to use please add a comment as to why this old dep needs to be used. In an upcoming version of https://github.com/envoyproxy/envoy/blob/master/tools/dependency/release_dates.py this will turn up as "aged".

**Review by moderation — CHANGES_REQUESTED** [review]
Please add comment as to why we need to pin to old commit

File: bazel/repository_locations.bzl
Comment (mathetake) [reply bazel/repository_locations.bzl:645]: https://github.com/bytecodealliance/wasmtime/tree/v0.21.0/crates/c-api here you can see the submodule's commit. I will add the comment

**Review by PiotrSikora — APPROVED** [review]

**Review by lizan — APPROVED** [review]

**Pull request merged**
